indexForBit optimization

jsign opened the pull request:
While getting to know the lib and playing around (beautiful data structure!), I ended up looking at some pprof were I noticed most of the allocs (count and size) for `Set` were in the bit calc regarding active slots for the bitwidth. Seems that was on the radar considering the existing comment.

I cleaned up the benchmark I was using to measure this (`BenchmarkSet`).
Then I tried to come up with another implementation which should have much less allocations.
Turns out it seems to make a big impact, here the `benchcmp`:
```
➜  go-hamt-ipld git:(setopt) ✗ benchcmp before.out after.out
benchmark                   old ns/op     new ns/op     delta
BenchmarkSet/1000/5-8       5434970       3767439       -30.68%
BenchmarkSet/1000/8-8       5334143       3846618       -27.89%
BenchmarkSet/10000/5-8      56021885      32351022      -42.25%
BenchmarkSet/10000/8-8      54762385      34444790      -37.10%
BenchmarkSet/100000/5-8     547704636     351883823     -35.75%
BenchmarkSet/100000/8-8     534793736     397958576     -25.59%

benchmark                   old allocs     new allocs     delta
BenchmarkSet/1000/5-8       57010          26353          -53.77%
BenchmarkSet/1000/8-8       50934          29709          -41.67%
BenchmarkSet/10000/5-8      560825         240196         -57.17%
BenchmarkSet/10000/8-8      511194         261228         -48.90%
BenchmarkSet/100000/5-8     5790539        2781380        -51.97%
BenchmarkSet/100000/8-8     5020547        3079226        -38.67%

benchmark                   old bytes     new bytes     delta
BenchmarkSet/1000/5-8       1803141       1151062       -36.16%
BenchmarkSet/1000/8-8       1937767       1370267       -29.29%
BenchmarkSet/10000/5-8      17631641      10604518      -39.86%
BenchmarkSet/10000/8-8      19236994      12051757      -37.35%
BenchmarkSet/100000/5-8     184884804     120835981     -34.64%
BenchmarkSet/100000/8-8     192682957     143161872     -25.70%
```
So that seems to indicate:
- ns/op -33%~
- allocs -50%~
- bytes -33%~

I kept both old and new impl in this branch if anyone wants to double-check easily renaming funcs and playing.  If it turns out this gets approved, I'll delete/rebase the old one for the final merging.

To be more sure about correctness, I did an extra check (not seen in this PR) that calls both old and new funcs and panics if return different results; it didn't panic for a `go test ./...` run nor with my added benchmark.

Stebalien commented on uhamt.go:
```diff
@@ -27,3 +27,17 @@ func popCount(i *big.Int) int {
+	var x uint64
+	var count, i int
+	w := n.Bitfield.Bits()
+	for x = uint64(bp); x > 64 && i < len(w); x -= 64 {
```
This needs to operate over uint and use `bits.UintSize` (not 64). Otherwise, it won't work on 32bit operating systems.

jsign replied:
Right!, the `Word` type of `BigInt` is a `uint`. Pushed.

Stebalien commented:
Hm. Nevermind, my library assumes that the array length is a multiple of 8.

Stebalien commented on uhamt.go:
```diff
@@ -27,3 +27,17 @@ func popCount(i *big.Int) int {
+	var count, i int
+	w := n.Bitfield.Bits()
+	for x = uint(bp); x > bits.UintSize && i < len(w); x -= bits.UintSize {
+		count += bits.OnesCount64(uint64(w[i]))
```
```suggestion
		count += bits.OnesCount(w[i])
```

Stebalien commented on uhamt.go:
```diff
@@ -27,3 +27,17 @@ func popCount(i *big.Int) int {
+	if i == len(w) {
+		return count
+	}
+	return count + bits.OnesCount64(uint64(w[i])&((1<<x)-1))
```
```suggestion
	return count + bits.OnesCount(w[i]&((1<<x)-1))
```

jsign commented:
> Actually, could you add a test case that extracts the current definition of this function into a test case and compares it against the new one? That way, we avoid any regressions.

@Stebalien , I refactored the code to make the test more narrow. Added two tests. 

A randomized one on 32 bytes `BigInt`s which should cover the bitwidth=8. Each case is fully verified in the 256bits that could be asked for. Challenging that test in the `i == len(w)` case (I changed the early return to be +1), it seemed that the randomized one didn't catch up that. Mostly because the randomized fashion + so many bits to cover, make that case a low chance.

Then I created simply a linear one that would be more covering on border cases, but on smaller bit lengths. In this test, that `i == len(w)` case broken-code fails instantly since the linear generation guarantees to cover it.

Kept both since wouldn't hurt.

Stebalien commented:
Awesome! Thanks!
